apps/emqx_message_validation/src/emqx_message_validation_schema.erl

```diff
@@ -0,0 +1,226 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2024 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%--------------------------------------------------------------------
+-module(emqx_message_validation_schema).
+
+-include_lib("typerefl/include/types.hrl").
+-include_lib("hocon/include/hoconsc.hrl").
+
+%% `hocon_schema' API
+-export([
+    namespace/0,
+    roots/0,
+    fields/1
+]).
+
+%% `minirest_trails' API
+-export([
+    api_schema/1
+]).
+
+-export([validate_name/1]).
+
+%%------------------------------------------------------------------------------
+%% Type declarations
+%%------------------------------------------------------------------------------
+
+%%------------------------------------------------------------------------------
+%% `hocon_schema' API
+%%------------------------------------------------------------------------------
+
+namespace() -> message_validation.
+
+roots() ->
+    [{message_validation, mk(ref(message_validation), #{importance => ?IMPORTANCE_HIDDEN})}].
+
+fields(message_validation) ->
+    [
+        {validations,
+            mk(
+                hoconsc:array(ref(validation)),
+                #{
+                    default => [],
+                    desc => ?DESC("validations"),
+                    validator => fun validate_unique_names/1
+                }
+            )}
+    ];
+fields(validation) ->
+    [
+        {tags, emqx_schema:tags_schema()},
+        {description, emqx_schema:description_schema()},
+        {enable, mk(boolean(), #{desc => ?DESC("config_enable"), default => true})},
+        {name,
+            mk(
+                binary(),
+                #{
+                    required => true,
+                    validator => fun validate_name/1,
+                    desc => ?DESC("name")
+                }
+            )},
+        {topics,
+            mk(
+                hoconsc:union([binary(), hoconsc:array(binary())]),
+                #{
+                    desc => ?DESC("topics"),
+                    converter => fun ensure_array/2,
+                    required => true
+                }
+            )},
+        {strategy,
+            mk(
+                hoconsc:enum([any_pass, all_pass]),
+                #{desc => ?DESC("strategy"), required => true}
+            )},
+        {failure_action,
+            mk(
+                hoconsc:enum([drop, disconnect]),
+                #{desc => ?DESC("failure_action"), required => true}
+            )},
+        {log_failure_at,
```

Comment:
in order to be prepared for future requirements such as log template etc.
guowei suggested to change this config to `log_failure { level = ... }`
so it's easier to add `log_failure {level = debug, template = ... }`

Reply:
:heavy_check_mark: